### DISCUSSION TITLE
runq_overload alert everyday

### DISCUSSION THREAD

**Discussion by chaymankala** [emqx/emqx#13188] [Q&A]
I have around 33,000 clients to my EMQX 5.5 cluster with 2 nodes.

Average messages received per sec ~ 200-300
Average messages sent per sec ~ 100-200

Nodes usage:
RAM: ~1.75/4GB (~45%)
CPU: ~20% (2 vCPUs combined)

I get this alert everyday for atleast 5-7 times
runq_overload:
`VM is overloaded on node: '<mynode1>@<ip>': 196`
The overload count is random sometimes its around 35 sometimes its as high as 500-600

I also followed everything mentioned in EMQX Tuning Guide (https://docs.emqx.com/en/emqx/v5.0/performance/tune.html)

I could not find any solution to this, please help me

**Comment by zmstone** [comment]
Hi
This alarm is triggered when Eralng VM scheduler is unable to schedule runnable processes in time.
Could you provide more information like
1. which authentication/authorization mechanisms are used
2. rule-engine / json data processing ?
3. data integration ?
4. message size, and traffic pattern? e.g. fan-in, fan-out etc.

**Comment by chaymankala** [comment]
Sure,
1. which authentication/authorization mechanisms are used
We are using MongoDB for both authentication and authorization. For authorization we are also caching data(32 items) for 1 hour
2. rule-engine / json data processing ?
We are not using any rule engine/flow designer
3. data integration ?
What do you mean by data integration?
4. message size, and traffic pattern? e.g. fan-in, fan-out etc.
avg. message size: 100-150 bytes
traffic pattern: almost constant, no sudden spikes 
Average messages received per sec ~ 300-400
Average messages sent per sec ~ 150-200

Average Authentication hit rate: 6-7times/sec
Average Authorization hit rate: 75-85 times/sec

**Reply by zmstone** [reply]
"3 data integration" means the data sink for "2 rule-engine" output, so this question is already answered. thanks.

I would recommend add 2 more vCPUs for each pod/host.

**Reply by chaymankala** [reply]
what configuration would you recommend for my use case if my clients go up to 100k?
And this issue wasn't there when we were using EMQX 4, are there any issues with v5 which is causing slow performance?

PS: Is there any option in EMQX to use more threads in CPU? because my CPU utilisation is already very low

**Reply by zmstone** [reply]
@chaymankala I just got reminded about a process leak bug in v5 before 5.7.0.
Maybe you should consider upgrading.

**Reply by zmstone** [reply]
more context: https://github.com/emqx/emqx/pull/13206
you can run this command to check if leak is happening in your nodes (the command should most of time print `0`):
```
emqx eval 'length(lists:filter(fun(Pid) -> case process_info(Pid,current_function) of {current_function,{qlc,wait_for_request,3}} -> true; _ -> false end end, processes())).'
```

**Reply by chaymankala** [reply]
Im sorry for the late reply, I ran the command few times its printing 0, does Zero mean it has process leak bug or not?
But anyway, im getting runq_overload only for the duration of 10-15 secs for 5-6 times day.
I will try upgrading to 5.7.0 in my dev server and will perform some tests and will let you know.
Thank you.

**Reply by chaymankala** [reply]
Update:
I tried upgrading to 5.7.0 and tested, I still have the issue. If i disable mongo authn and authz there are no alarms, I guess this is related to MongoDB connection. Currently Im caching authz for 12 hrs and current rate is ~75 times/sec and authn is ~8 times/sec
I indexed all the fields where its queried and I enabled readSecondary, is there anything I can do more to optimise this?

**Reply by qzhuyan** [reply]
speak of `VM is overloaded on node: '<mynode1>@<ip>': 196`, Do you see CPU spikes at the moment of the alarm? 

If so, you should consider add more VCPUs.
If not, maybe you VM provider is limiting you at IO or network, check the other metrics (network, disk IO).

**Reply by chaymankala** [reply]
Everything is normal, I checked this in multiple instances, posting the screenshot of one incident for reference (occurred at 11:25AM today IST)

<img width="1327" alt="Screenshot 2024-06-12 at 3 16 36 PM" src="https://github.com/emqx/emqx/assets/26591158/bd30cf5f-d9a8-4c2b-b2fd-e09aaa00211d">
<img width="1061" alt="Screenshot 2024-06-12 at 3 16 20 PM" src="https://github.com/emqx/emqx/assets/26591158/281baa47-ff6e-42e6-bf2d-cec15a885e96">

**Reply by chaymankala** [reply]
One more bug too, the above screenshots are from EMQX NODE 1 (2 Nodes in our cluster), 
But in NODE 2, there is the alarm since a week and its not shutting off,
I ran this command in NODE2 and the output is 0
emqx eval 'length(lists:filter(fun(Pid) -> case process_info(Pid,current_function) of {current_function,{qlc,wait_for_request,3}} -> true; _ -> false end end, processes())).'

Attaching screenshot
<img width="1332" alt="Screenshot 2024-06-12 at 3 23 34 PM" src="https://github.com/emqx/emqx/assets/26591158/4dd93237-b390-430f-a041-190ddeae4723">

**Reply by qzhuyan** [reply]
```
emqx eval 'length(lists:filter(fun(Pid) -> case process_info(Pid,current_function) of {current_function,{qlc,wait_for_request,3}} -> true; _ -> false end end, processes())).'
```
this command only checks the specific issue for the https://github.com/emqx/emqx/pull/13206

could you run following commands on NODE2?
```emqx eval "emqx_olp:is_overloaded()"```


what is the resolution of your CPU metrics?  You mentioned "duration of 10-15 secs", that means you need smaller resolution to catch it in your metrics history. also have you check the network metrics?

**Reply by chaymankala** [reply]
Duration of 10-15secs is given in EMQX dashboard (Alarms), Im attaching screenshot again (see in the last column).

emqx eval "emqx_olp:is_overloaded() is returning false in NODE2
<img width="1327" alt="Screenshot 2024-06-12 at 3 16 36 PM" src="https://github.com/emqx/emqx/assets/26591158/b9cf0d20-d3d8-4872-a2ba-91c437611198">

**Reply by qzhuyan** [reply]
ok. so it is time to scale up your node with more cores.

**Reply by chaymankala** [reply]
I tried scaling up in a staging server and the issue is same even with 4 core CPU. This is happening only when I enable authn and authz with MongoDB. I guess there is some serious issue with authn/authz MongoDB plugins, please try to look into this. You can reproduce easily by running a benchmark tool with around 30k clients with authn/authz.
Hope this will be fixed in coming versions.

**Reply by chaymankala** [reply]
And also please look into this issue why the alarms get stuck like this (Please refer to screenshot), these alarms wont go until I restart nodes.
even this command  (emqx eval "emqx_olp:is_overloaded()")  is return false in both nodes, looks like both of them are false alarms

<img width="1339" alt="Screenshot 2024-06-20 at 2 23 33 PM" src="https://github.com/emqx/emqx/assets/26591158/b54e79f1-f97f-4fd4-9e02-bc1f6041c5ae">

**Reply by qzhuyan** [reply]
OK. Could raise issues separately and with logs attached for speeding up the troubleshooting?

**Reply by chaymankala** [reply]
Sure I will raise issues,
Thank you
